core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c

```diff
@@ -969,9 +988,26 @@ wasi_poll_oneoff(wasm_exec_env_t exec_env, const wasi_subscription_t *in,
         || !validate_native_addr(nevents_app, sizeof(uint32)))
         return (wasi_errno_t)-1;
 
-    err = wasmtime_ssp_poll_oneoff(curfds, in, out, nsubscriptions, &nevents);
-    if (err)
-        return err;
+    int elapsed = 0;
+    /* in seconds */
+    const int timeout = get_timeout_for_poll_oneoff(in), time_quant = 1;
+    __wasi_subscription_t in_copy[nsubscriptions];
+    memcpy(in_copy, in, nsubscriptions * sizeof(wasi_subscription_t));
+
+    while (1) {
+        elapsed += time_quant;
+        if (timeout != -1 && elapsed > timeout)
+            break;
+
+        in_copy[0].u.u.clock.timeout = time_quant * 1e9;
+        err = wasmtime_ssp_poll_oneoff(curfds, in_copy, out, nsubscriptions,
+                                       &nevents);
+        if (err)
+            return err;
+
+        if (wasm_runtime_get_exception(module_inst) || nevents_app >= 0)
```

Comment:
isn't the condition `nevents_app >= 0` always true?

Reply:
was a bit unsure of this here, should the interrupt be dependent on the number of events and just rely on the exception occurrence?

Reply:
We should stop the loop when the event happened (we shouldn't call `poll_oneof` again, it would be incorrect behavior). Having said that, when there's an event, we shouldn't return `EINTR` error code.

Reply:
Agree with @loganek, should stop the loop here when the event(s) occured.
And here we should use `nevents > 0` instead of `nevents_app >= 0`, nevents_app is the address pointing to some place of the linear memory space, it will always `>= 0` and is a valid address since it is checked by L988.
nevents is the returned count of events occured